Glossary: Fix #380

tdelmas opened the pull request:
A glossary of main terms related to Let's Encrypt and PKI

As the objective is to help people with not enough knowledge of the PKI I tried to include most terms that can be found in the Let's Encrypt website.

I assumed a minimal knowledge, so terms such as DNS are not present (It may be a mistake, please tell me if you felt differently)

Some terms with few relation to Let's Encrypt or PKI may be present because:
- They are used in other definitions
- Helps to contrast some definitions (such as "Web server"/"Web client" to contrast with "ACME server"/"ACME client")

Please tell me if you feel some terms should be removed.

I may have over engineered the definition macro, but I felt it was important for accessibility (especially for screen readers, to know the language and if it's an acronym)
And if somebody have more knowledge about accessibility please step in!

Terms that could be added: JWS, API, SSL/TLS, QUIC, SPDY, TCP, UDP, DNS, HTTP, HTTP/2, HTTPS, IP, IPv4, IPv6, SSH, NPN, ALPN, CBC, GCM, AES, SHA, MD5, Forward Secrecy, hash function, Cipher Suites.

Some definitions use a small extract from their Wikipedia article or are inspired by it. It's only short extract and there is a link to the Wikipedia in the end of the definition, by maybe I should add a small text in the end of the page indicating it (with full license information), or rewrite them to not be so close of the Wikipedia article?

Changes other than `glossary.md`:
- I had to add `beforeColon` in `config.toml` for i18n.
- `glossary.css` to add an highlight on the defined word when switching from a definition to another by clicking on an internal link (useful to see when the definition starts when clicking on a word to see it's definition)
- `glossary.js` to add on internal links pointing to another definition, that definition on the title (useful to see the definition of a word without having to click a word to see it's definition)
- `docs/_index.md` to add the link to `glossary.md`
- ` layouts/shortcodes/def.html` for the `def` [hugo shortcode](https://gohugo.io/content-management/shortcodes/)

Here is the paste of the rendered text as an example:


 
**ACME Client**: A software capable to communicate with an ACME server to ask for a [certificate](#def-leaf "Leaf certificate (end-user certificate): A certificate signed by a trusted intermediate, valid for a set of domains."). 
**ACME Server**: An ACME-compatible server capable to generate [certificates](#def-leaf "Leaf certificate (end-user certificate): A certificate signed by a trusted intermediate, valid for a set of domains."). Let’s Encrypt software, [Boudler](boudler), is ACME-compatible. [Boulder divergences from ACME](https://github.com/letsencrypt/boulder/blob/master/docs/acme-divergences.md) 
**Automatic Certificate Management Environment** (ACME): The protocol implemented by [Let’s Encrypt](#def-LE "Let's Encrypt (LE): The Certificate Authority controlled by ISRG."). Softwares compatibles with that protocol can use it to communicate with Let’s Encrypt to ask for a [certificate](#def-leaf "Leaf certificate (end-user certificate): A certificate signed by a trusted intermediate, valid for a set of domains."). [ACME draft 16](https://tools.ietf.org/html/draft-ietf-acme-acme-16) - [Wikipedia](https://en.wikipedia.org/wiki/Automated_Certificate_Management_Environment) 
**Boudler**: The software implementing ACME, developed and used by [Let’s Encrypt](#def-LE "Let's Encrypt (LE): The Certificate Authority controlled by ISRG."). [GitHub](https://github.com/letsencrypt/boulder)

tdelmas commented on content/fr/docs/glossary.md:
```diff
@@ -0,0 +1,138 @@
+{{% def id="ACME-client" name="Client ACME" %}} a software capable to communicate with an ACME server to ask for a [certificate](#def-leaf). {{% /def %}}
+
+
+{{% def id="ACME-server" name="Serveur ACME" %}} an ACME-compatble server capable to generate [certificates](#def-leaf). Let's Encrypt software, [Boudler](boudler), is ACME-compatible. [Boulder divergences from ACME](https://github.com/letsencrypt/boulder/blob/master/docs/acme-divergences.md) {{% /def %}}
```
Thanks!

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,127 @@
+
+{{% def id="ACME" name="Automatic Certificate Management Environment" abbr="ACME" %}} The protocol implemented by [Let's Encrypt](#def-LE). Softwares compatibles with that protocol can use it to communicate with Let's Encrypt to ask for a [certificate](#def-leaf). [ACME draft 16](https://tools.ietf.org/html/draft-ietf-acme-acme-16) - [Wikipedia](https://en.wikipedia.org/wiki/Automated_Certificate ...
+
+{{% def id="boudler" name="Boudler" %}} The software implementing ACME, developed and used by [Let's Encrypt](#def-LE). [GitHub](https://github.com/letsencrypt/boulder) {{% /def %}}
```
Thanks!

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,127 @@
+{{% def id="ACME-client" name="ACME Client" %}} A software capable to communicate with an ACME server to ask for a [certificate](#def-leaf). {{% /def %}}
+
+
+{{% def id="ACME-server" name="ACME Server" %}} An ACME-compatible server capable to generate [certificates](#def-leaf). Let's Encrypt software, [Boudler](boudler), is ACME-compatible. [Boulder divergences from ACME](https://github.com/letsencrypt/boulder/blob/master/docs/acme-divergences.md) {{% /def %}}
```
Thanks!

jsha commented:
This is great work, thanks @tdelmas ! I made some edits fixing style and a few technical points. I tried pushing those directly to your branch to save work, but I appear to have merged in some unrelated history from the master branch as well, creating diff that appears to include unrelated stuff. Sorry about that! I'm going to try a local rebase and push to fix it.

jsha commented:
Alright, I believe that fixed it. Please let me know what you think of my edits; none are fixed in stone, but this seemed like the easiest way to collaborate. There are a couple more I'd like to make, that I'll push either later tonight or tomorrow:

- I'd like to change the "web client" terminology to use "User Agent" which I think is more standard, and is harder to confuse with "ACME client."
- I'd like to remove the OID listings. I appreciate all the work you put into listing them, but I think the target audience for this glossary is server administrators, and OID numbers are mainly useful to people implementing either CAs or client software, and those groups can reference the docs. And I think including the long strings of digits is likely to increase the impression of some novice users that PKI is arcane.
- I'll add a definition of Web PKI. I added links to it but forgot to document it!

BTW, I particularly love the behavior you added where hovering on a definition link gives a tooltip. That's a great way to get a quick peek at a definition while you're reading, without losing your place!

jsha commented:
Okay, I've pushed those. There was one other tweak I wanted to get your thoughts on: For some of these terms, the acronym is the more typical way to write it, and the expansion is more rare. So for those I think it makes sense to put the acronym first. For instance, "DANE", "OCSP" and "TLS" would fit into that category, but probably not "Certificate Transparency."

What would you think of just swapping the "name" and "abbr" fields in that case? Or we could change the "abbr" field to "alt" to capture the fact that it wouldn't always be an abbreviation?


E.g. it might look like:

TLS (Transportation Security Layer): ... definition ...
DANE (DNSSEC-based Authentication of Named Entities): ... definition ...

tdelmas commented:
@jsha Thank you! I should have time to write an answer tonight. I put back the "WIP" in the title

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="CPS" name="Certification Practice Statement" abbr="CPS" %}} A statement of the practices that a certification authority employs in issuing, managing, revoking, and renewing or re-keying certificates. [ISRG Certification Practice Statement](https://letsencrypt.org/repository/#isrg-certification-practice-statement) - [RFC 3647 section 3.4](https://tools.ietf.org/html/rfc3647#section-3.4 ...
+
+{{% def id="critical" name="Critical" %}} A certificate may contain [extensions](#def-extension) marked "critical." This means that software must reject that certificate unless the software understands how to process that extension. This makes it possible to introduce new extensions that are important for security without creating risks for older software. {{% /def %}}
```
"Critical" seams too generic for me, I think "Critical extension" is better

```suggestion
{{% def id="critical" name="Critical extension" %}} A certificate may contain [extensions](#def-extension) marked "critical." This means that software must reject that certificate unless the software understands how to process that extension. This makes it possible to introduce new extensions that are important for security without creating risks for older software. {{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="EV" name="Extended Validation" abbr="EV" %}} A type of certificate validation for which the [CA](#def-CA) has verified the legal entity controlling the website. They contain information about that entity. Controls from the [CA](#def-CA) are more strict than for [OV](#def-OV) certificates. [Let's Encrypt](#def-LE) doesn't offer EV certificates. [Wikipedia](https://en.wikipedia.org/wiki ...
+
+{{% def id="extension" name="Extension" %}} In certificates, most fields are defined by extensions. For instance, [Subject Alternative Names] and [AIA](#def-AIA) are extensions. The extension mechanism allows creating new fields that were not part of the original [X.509](#def-X509) standard. {{% /def %}}
```
"Extension" seams too generic for me, I think "Certificate extension" is better

```suggestion
{{% def id="extension" name="Certificate extension" %}} In certificates, most fields are defined by extensions. For instance, [Subject Alternative Names] and [AIA](#def-AIA) are extensions. The extension mechanism allows creating new fields that were not part of the original [X.509](#def-X509) standard. {{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="ISRG" name="Internet Security Research Group" abbr="ISRG" %}} The organization behind [Let's Encrypt](#def-LE): https://www.abetterinternet.org/about/. [Wikipedia](https://en.wikipedia.org/wiki/Internet_Security_Research_Group) {{% /def %}}
+
+{{% def id="Issuer" name="Issuer" %}} A certificate field describing which certificate signed the current one. For instance, the Issuer field of a Let's Encrypt end-entity certificate might be "Issuer: C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3". It commonly contains fields like [Common Name](#def-CN), Country, and Organization. The Issuer field always matches some certificate's [ ...
```
"Issuer" seams too generic for me, I think "Certificate issuer" is better

```suggestion
{{% def id="issuer" name="Certificate issuer" %}} A certificate field describing which certificate signed the current one. For instance, the Issuer field of a Let's Encrypt end-entity certificate might be "Issuer: C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3". It commonly contains fields like [Common Name](#def-CN), Country, and Organization. The Issuer field always matches some certificate's [Subject](#def-Subject) field. For [self-signed](#def-self-signed) certificates like [roots](#def-root), the Issuer is the same as the Subject. The term "issuer" may also be used to indicate a certificate that issues other certificates (an [intermediate](#def-intermediate) or root), or an organization that issues certificates.{{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="staging" name="Staging" %}} [Let's Encrypt](#def-LE) provides a staging API to test certificate request without impacting rate limits. Certificates generated by the staging environment are *not* publicly trusted. The staging environment should be used for testing, debugging, and ACME client development purposes. https://letsencrypt.org/docs/staging-environment/ {{% /def %}}
+
+{{% def id="Subject" name="Subject" %}} A certificate field indicating what a certificate is about. This commonly contains fields like [Common Name](#def-CN), Country, and Organization. {{% /def %}}
```
"Subject" seams too generic for me, I think "Certificate subject" is better

```suggestion
{{% def id="subject" name="Certificate subject" %}} A certificate field indicating what a certificate is about. This commonly contains fields like [Common Name](#def-CN), Country, and Organization. {{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="IDNA" name="Internationalized Domain Names for Applications" abbr="IDNA" %}} See [internationalized domain name](#def-IDN). {{% /def %}}
+
+{{% def id="IDN" name="Internationalized Domain Name" abbr="IDN" %}} Domain name with characters others than `a` to `z`, `0` to `9` and the hyphen (`-`). They can for example contain Arabic, Chinese, Cyrillic, Tamil, Hebrew or Latin alphabet-based characters with diacritics or ligatures. The encoded representation of an IDN domains starts with `xn--`. IDNs are supported by [Let's Encrypt](#def-LE ...
```
missing "-" between links

```suggestion
{{% def id="IDN" name="Internationalized Domain Name" abbr="IDN" %}} Domain name with characters others than `a` to `z`, `0` to `9` and the hyphen (`-`). They can for example contain Arabic, Chinese, Cyrillic, Tamil, Hebrew or Latin alphabet-based characters with diacritics or ligatures. The encoded representation of an IDN domains starts with `xn--`. IDNs are supported by [Let's Encrypt](#def-LE): https://letsencrypt.org/2016/10/21/introducing-idn-support.html. [Wikipedia](https://en.wikipedia.org/wiki/Internationalized_domain_name) - [RFC 5890](https://tools.ietf.org/html/rfc5890) - [RFC 5891](https://tools.ietf.org/html/rfc5891) {{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="Key-pair" name="Key-pair" %}} A combination of a private key and public key used to sign or encrypt. The public key is commonly embedded in a certificate, while the private key is stored on its own and should be kept secret. A key pair can be used to encrypt and decrypt, to sign and verify data, or to negotiate secondary keys, depending on the application. [Wikipedia](https://en.wikip ...
+
+{{% def id="leaf" name="Leaf certificate (end-entity certificate)" %}} Most commonly, a certificate signed by an [intermediate](#def-intermediate), valid for a set of domains and not able to sign other certificates. This is the type of certificate that [ACME clients](#def-ACME-client) request, and that [web servers](#def-web-server) use. [Wikipedia](https://en.wikipedia.org/wiki/Public_key_certif ...
```
"end-entity certificate" is the correct term but "end-user certificate" seams commonly used too, maybe we can have both? (I'm not convinced it's better than only "end-entity", but...)

```suggestion
{{% def id="leaf" name="Leaf certificate (end-entity/end-user certificate)" %}} Most commonly, a certificate signed by an [intermediate](#def-intermediate), valid for a set of domains and not able to sign other certificates. This is the type of certificate that [ACME clients](#def-ACME-client) request, and that [web servers](#def-web-server) use. [Wikipedia](https://en.wikipedia.org/wiki/Public_key_certificate#End-entity_or_leaf_certificate) {{% /def %}}
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="certificate" name="Certificate" %}} A file in a [particular format](#def-X509) that contains a public key and other data describing when to use that public key. The most common kind of certificate is a [leaf certificate](#def-leaf). There are also [intermediate](#def-intermediate) and [root](#def-root) certificates. {{% /def %}}
+
+{{% def id="CA" name="Certificate Authority" abbr="CA" %}} An organization that issues [certificates](#def-leaf). [Let's Encrypt](#def-LE), Sectigo, and DigiCert are Certificate Authorities. [Wikipedia](https://en.wikipedia.org/wiki/Certificate_authority) {{% /def %}}
```
It's already defined before :-) And the previous one is - I think - in the correct position alphabetically, no?
```suggestion
```

tdelmas commented on content/en/docs/glossary.md:
```diff
@@ -0,0 +1,187 @@
+
+{{% def id="CNAME" name="Canonical Name record" abbr="CNAME" %}} A DNS entry which maps one domain name to another, referred to as the Canonical Name. [Wikipedia](https://en.wikipedia.org/wiki/CNAME_record) {{% /def %}}
+
+{{% def id="CA" name="Certificate Authority" abbr="CA" %}} An organization that issues [certificates](#def-leaf). [Let's Encrypt](#def-LE) and [IdenTrust](#def-IdenTrust) are Certificate Authorities. [Wikipedia](https://en.wikipedia.org/wiki/Certificate_authority) {{% /def %}}
```
Merge of the other definition of "CA"
```suggestion
{{% def id="CA" name="Certificate Authority" abbr="CA" %}} An organization that issues [certificates](#def-leaf). [Let's Encrypt](#def-LE), [IdenTrust](#def-IdenTrust), Sectigo, and DigiCert are Certificate Authorities. [Wikipedia](https://en.wikipedia.org/wiki/Certificate_authority) {{% /def %}}
```

tdelmas commented:
> What would you think of just swapping the "name" and "abbr" fields in that case? Or we could change the "abbr" field to "alt" to capture the fact that it wouldn't always be an abbreviation?

I think an option for the def macro called "abbr_first" should do it. (If it's not an abbreviation, it should be in the name, so the order is the one we choose for the content of the name, no?)
Can I update this PR in that direction?

Another point: to be consistent, I think IDs should be upper-case for abbreviations and lower-case for other, I made the mistake for "Key-pair", can I fix it to be "key-pair"?

tdelmas commented:
Hi @jsha, here are my changes: https://github.com/tdelmas/website/compare/730d40e4c7...3e83f4ee4c

I've merge my suggestions with your comments and the key-pair id

I've added the abbr_first argument and used it for SSL, TLS, OCSP, ECDSA, EdDSA

I've added an entry "End-user certificate" that point to "Leaf certificate (end-entity certificate)" (I didn't mentioned end-user in the leaf definition because it's not the "correct" term, but feel free to do it). I've seen it a lot, for example a google search for the exact term "end-entity certificate"+"certificate authority" gives 9 200 results https://www.google.com/search?q=%22end-entity+certificate%22+%22certificate+authority%22 when "end-user certificate"+"certificate authority" gives 17 400 results https://www.google.com/search?q=%22end-user+certificate%22+%22certificate+authority%22 
Examples:
https://ssl.comodo.com/articles/understanding-an-ssl-certificate-chain.php
https://support.dnsimple.com/articles/what-is-ssl-certificate-chain/

Another point: I feel we should talk about "web client" and not "browser" in the "OCSP Must-Staple" and  "OCSP stapling" definitions, should I fix it ? (it may be fixed latter too)

jsha commented:
>  I've seen it a lot, for example a google search for the exact term "end-entity certificate"+"certificate authority" gives 9 200 results

Google's results counts are very approximate, especially for complex queries (e.g. with quotes). I don't think we can deduce popularity from them. I note that the first results for "end-user certificate" by itself is https://en.wikipedia.org/wiki/End-user_certificate, which suggests that the term is about the arms trade.

I think when you search for ["end-user certificate" "certificate authority"] you are seeing sporadic cases of documentation authors making a mistake. For instance, the top result is this DNSimple page, which says "end-user": https://support.dnsimple.com/articles/what-is-ssl-certificate-chain/. But we can't conclude that DNSimple prefers "end-user." Look at these two searches:

 - [[site:dnsimple.com "end-user certificate"]](https://www.google.com/search?q=site%3Adnsimple.com+%22end-user+certificate%22&oq=site%3Adnsimple.com+%22end-user+certificate%22&aqs=chrome..69i57j69i58.6855j1j7&sourceid=chrome&ie=UTF-8) (1 result)
 - [[site:dnsimple.com "end-entity certificate"]](https://www.google.com/search?ei=prszXNHVO9m70PEP8-if0AE&q=site%3Adnsimple.com+%22end-entity+certificate%22&oq=site%3Adnsimple.com+%22end-entity+certificate%22&gs_l=psy-ab.3...14202.14453..14772...0.0..0.41.80.2......0....1..gws-wiz.......0i71.Mhr0a45aWiM) (9 results, but Google strips the quotes)

I'd prefer to leave out "end-user certificate." I believe it's a term of art in another field, and usually a mistake when written in the PKI world, and I would prefer that we not be responsible for introducing another confusing synonym for end-entity certificate (or leaf certificate or server certificate).

> Another point: I feel we should talk about "web client" and not "browser" in the "OCSP Must-Staple" and "OCSP stapling" definitions, should I fix it ? (it may be fixed latter too)

I slightly prefer to talk about "browser" whenever possible. I think it favors clarity over generality. But I don't feel too strongly about this.

I'm doing what will probably be my last round of review before merging. I noticed a small bug in the JS: When I hover over "issuer URI" in the definition "Authority Information Access (AIA): A certificate extension used to indicate to user agents how to obtain information about the issuer of the certificate. It typically specifies the OCSP URI and the issuer URI.", the tooltip is for "Certificate authority" (#def-CA) rather "Certificate Authority Issuer" (#def-CAI).

jsha commented:
> Fix also fixed a few links, but I didn't succeed to reproduce your title bug... I see the correct (for CAI) tooltip over "issuer URI".

I also can't reproduce this. Cool!

Merging now.
